From dcaee6fd0d7a9e249decea610ab40fe712f07ef5 Mon Sep 17 00:00:00 2001
From: Jordan Hendricks <jordan.hendricks@joyent.com>
Date: Tue, 20 Mar 2018 00:39:41 +0000
Subject: [PATCH] MANTA-3410 Audit job reports missing MPU parts MANTA-3608
 update mola node-verror dependency MANTA-3613 update mola nodeunit dependency

---
 lib/auditor.js | 40 +++++++++++++++++++++++++++++++++++-----
 package.json   |  4 ++--
 2 files changed, 37 insertions(+), 7 deletions(-)

diff --git a/lib/auditor.js b/lib/auditor.js
index 7b592f9..b29ba9a 100644
--- a/lib/auditor.js
+++ b/lib/auditor.js
@@ -1,18 +1,33 @@
 /*
- * This Source Code Form is subject to the terms of the Mozilla Public
+  * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
-var util = require('util');
-var events = require('events');
+var assert = require('assert-plus');
 var carrier = require('carrier');
+var events = require('events');
+var util = require('util');
 
 
+///--- Helpers
+
+/*
+ * Returns true if the given key is under the directory structure
+ * "/:accountUuid/uploads/", and false otherwise.
+ */
+function uploadsDirChild(key) {
+        assert.string(key, 'key');
+
+        /* JSSTYLED */
+        var uploadsRegex = /^\/[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\/uploads\//;
+        return (key.match(uploadsRegex));
+}
+
 
 ///--- API
 
@@ -62,7 +77,22 @@ function Auditor(opts, listener) {
                 } else {
                         if ((currMako.objectId !== objectId) ||
                             (currMako.storageNodes.indexOf(storageId) === -1)) {
-                                self.emit('problem', line);
+
+                            if (parts.length >= 4) {
+                                var key = parts[3];
+
+                               /*
+                                * Committed, non-garbage collected multipart
+                                * uploads may have parts with no backing mako
+                                * file, which are removed during normal MPU
+                                * commit operations.
+                                */
+                                if (uploadsDirChild(key)) {
+                                        return;
+                                }
+                            }
+
+                            self.emit('problem', line);
                         }
                 }
         });
diff --git a/package.json b/package.json
index 8d477f7..70b65c0 100644
--- a/package.json
+++ b/package.json
@@ -23,11 +23,11 @@
                 "posix-getopt": "1.0.0",
                 "sprintf-js": "0.0.7",
                 "vasync": "1.6.1",
-                "verror": "^1.9.0",
+                "verror": "^1.10.0",
                 "vstream": "0.1.0"
         },
         "devDependencies": {
-                "nodeunit": "0.7.4"
+                "nodeunit": "0.9.1"
         },
         "sdcDependencies": {
                 "config-agent": ">=1.2.0"
-- 
2.21.0

